config/opal_setup_java.m4

```diff
@@ -93,18 +93,22 @@ AC_DEFUN([OPAL_SETUP_JAVA],[
         # hard-code a few of the common ones so that users don't have to
         # specify --with-java-<foo>=LONG_ANNOYING_DIRECTORY.
         AS_IF([test -z "$with_jdk_bindir"],
-              [ # OS X Snow Leopard and Lion (10.6 and 10.7 -- did not
-                # check prior versions)
+              [ # OS X/macOS
                opal_java_found=0
-               AS_IF([test -x /usr/libexec/java_home],
-                    [opal_java_dir=`/usr/libexec/java_home`/include],
-                    [opal_java_dir=/System/Library/Frameworks/JavaVM.framework/Versions/Current/Headers])
-               AC_MSG_CHECKING([OSX locations])
-               AS_IF([test -d $opal_java_dir],
+               AS_IF([test -x /usr/libexec/java_home], 
+                     [opal_java_dir=`/usr/libexec/java_home`], 
+                     [opal_java_dir=/System/Library/Frameworks/JavaVM.framework/Versions/Current])
+               AC_MSG_CHECKING([OS X/macOS locations])
+               AS_IF([test -d $opal_java_dir], 
                      [AC_MSG_RESULT([found ($opal_java_dir)])
                       opal_java_found=1
-                      with_jdk_headers=$opal_java_dir
-                      with_jdk_bindir=/usr/bin],
+                      if test "$opal_java_dir" = /System/Library/Frameworks/JavaVM.framework/Versions/Current; then
```

Comment:
instead of testing the value of `$opal_java_dir`, could we check the existence of `$opal_java_dir/{include,bin}` ?

Reply:
@ggouaillardet:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Not unless you want to drop support for Apple's Java shims.  (I'm not sure what their status is, but Apple stopped shipping its own derivative Java distribution after Java…6, I believe, so odds are they've actually been deprecated since some time after that…?)  I noted in my toolpath analysis [above](https://github.com/open-mpi/ompi/pull/5015#issuecomment-379324639) that the header and executable directories under `$opal_java_dir` differ in naming scheme depending on what that variable's contents are, did I not?  In any case, I felt it safer to leave messing with that to the future.

Reply:
@RandomDSdevel First of all, many thanks for your patience.  This is a complex topic; forgive us for being so pedantic trying to understand it.

1. I would actually love to see much of https://github.com/open-mpi/ompi/pull/5015#issuecomment-379324639 turned into a comment in the .m4 file.  This is complex stuff; putting all that hard-won knowledge down as a comment in the code could save someone from making a mistake in the future while "fixing" something (but not having the background knowledge of why the code currently is the way it is).
1. I think what @ggouaillardet was suggesting was not incongruent with your analysis.  I think he was suggesting something like this (typed directly here in github / not tested -- YMMV):

```sh
if test -d "$opal_java_dir/Headers" && test -d "$opal_java_dir/Commands"; then
    with_jdk_headers="$opal_java_dir/Headers"
    with_jdk_bindir="$opal_java_dir/Commands"
elif test -d "$opal_java_dir/include" && test -d "$opal_java_dir/bin"; then
    with_jdk_headers="$opal_java_dir/include"
    with_jdk_bindir="$opal_java_dir/bin"
else
    AC_MSG_WARN([No recognized directory structure found under $opal_java_dir])
    AC_MSG_ERROR([Cannot continue])
fi
```

This simply prevents us from having the hard-coded ".../Current" directory name in 2 places.  It also is marginally more defensive in that it also checks for the presence of `include` and `bin` if the `Headers` and `Commands` checks fail.

Is there a reason that would drop support for Apple's Java shims?  It looks like the same logic to me -- is there a subtle implication that I'm missing?

Reply:
@jsquyres:  

> @RandomDSdevel First of all, many thanks for your patience. This is a complex topic; forgive us for being so pedantic trying to understand it.
>
> ⋮

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;You're welcome, and no problem.  

> ⋮
>
>1. I would actually love to see much of [#5015 (comment)](https://github.com/open-mpi/ompi/pull/5015#issuecomment-379324639) turned into a comment in the .m4 file. This is complex stuff; putting all that hard-won knowledge down as a comment in the code could save someone from making a mistake in the future while "fixing" something (but not having the background knowledge of why the code currently is the way it is).
>
> ⋮

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Agreed, but…that would be rather a lot of text, don't you think?  I could just put URIs pointing to the relevant discussion into such a comment instead; that would be a lot shorter.  

> ⋮
>
> 2. I think what @ggouaillardet was suggesting was not incongruent with your analysis. I think he was suggesting something like this (typed directly here in github / not tested -- YMMV):
>
>    ```
>    if test -d "$opal_java_dir/Headers" && test -d "$opal_java_dir/Commands"; then
>        with_jdk_headers="$opal_java_dir/Headers"
>        with_jdk_bindir="$opal_java_dir/Commands"
>    elif test -d "$opal_java_dir/include" && test -d "$opal_java_dir/bin"; then
>        with_jdk_headers="$opal_java_dir/include"
>        with_jdk_bindir="$opal_java_dir/bin"
>    else
>        AC_MSG_WARN([No recognized directory structure found under $opal_java_dir])
>        AC_MSG_ERROR([Cannot continue])
>    fi
>    ```
>
> This simply prevents us from having the hard-coded ".../Current" directory name in 2 places. It also is marginally more defensive in that it also checks for the presence of `include` and `bin` if the `Headers` and `Commands` checks fail.
>
> ⋮

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Ah, OK; that does seem a bit nicer, so I'll put that in place now.  

> ⋮
>
> Is there a reason that would drop support for Apple's Java shims? It looks like the same logic to me -- is there a subtle implication that I'm missing?

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sorry, I should have repeated some additional context from downstream in https://github.com/Homebrew/homebrew-core/issues/26009#issuecomment-377735620 here; here it is:  

> … (Per [an old message which I found on one of the OpenJDK mailing lists](http://mail.openjdk.java.net/pipermail/macosx-port-dev/2014-February/006388.html) via [this Ask Different answer](https://apple.stackexchange.com/a/192200/70614) when I was looking around to see what information I could jump off from to troubleshoot [a possible lead on this issue]:  
>
>> ⋮
>>
>> Do NOT remove any content in the JavaVM.framework. Those items are required by Java 7, 8, 9+ as well as Java SE 6. No modern version of OS X has a Java JDK inside there anyway.
>>
>> ⋮
>
> ⋮

However, if you run:  

```
…$ sudo -i
⋮ # Enter password…
…# find / -type f -perm +111 -exec file -p {} \; -exec head -c 1 /dev/zero \; | xargs -0 -n 1 sh -c 'echo "$@" | head -n 1' -- | grep -v x86_64 | grep -E Mach-O.+i386
⋮ # Wait for that to finish…
…# logout
```

(at least on my system, which has OS X v10.11.6 'El Capitan,' build 15G20015, installed,) then you'll notice that the resulting long list includes the following output:  

```
⋮
/System/Library/Frameworks/JavaVM.framework/Versions/A/Resources/.compatibility/libIASupport.jnilib: Mach-O dynamically linked shared library i386
⋮
```

There might also be some 32-bit components elsewhere on Macs upon which `/System/Library/Frameworks/JavaMV.framework` might additionally depend, but I haven't looked into what those might be.  (I also haven't asked anywhere to see if this might have changed in versions of OS X/macOS newer than the one I have installed on my system.)  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;What implications this factoid might have, if any, for the future of Apple's Java shims, I don't know, but all 32-bit OS X/macOS binaries currently have a limited lifespan, as noted in [this notice](https://developer.apple.com/news/?id=06282017a) on Apple's Developer portal:  

> ## 64-bit Requirement for Mac Apps
>
> June 28, 2017
>
> At WWDC 2017, we announced new apps submitted to the Mac App Store must support 64-bit starting January 2018, and Mac app updates and existing apps must support 64-bit starting June 2018. If you distribute your apps outside the Mac App Store, we highly recommend distributing 64-bit binaries to make sure your users can continue to run your apps on future versions of macOS. macOS High Sierra will be the last macOS release to support 32-bit apps without compromise.

Whether Apple will use this aa an excuse to remove their vestigial Java shims entirely or choose to update that lone 32-bit component inside it (and any the entire shim framework might depend on) to be 64-bit executables, I again don't know.  Relying solely on users only having instances of Oracle's official JDK installed on their system _might_, however, turn out to be a safer idea in the long run, but I'm deeming that outside the scope of this PR for now.

Reply:
@jsquyres:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I've addressed most of your review comments with a new revision of this PR's sole commit pushed to its staging branch on my Open MPI repository fork.  I left out adding an additional comment as you asked for now, though, pending any guidance you might have on what you might want it to look like (just something short pointing through one or more link vs. a copied-and-pasted variant of my giant wall of text from https://github.com/open-mpi/ompi/pull/5015#issuecomment-379324639.)  (I should also note that said wall of text was adapted from an e-mail that I posted to devel@lists.open-mpi.org and mentioned in my commit message.)